### PR TITLE
[protocol] split gas fee and burn base fee

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
1. spilt gas and burn base for transfer protocol
2. TODO: add other protocols

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by CoderZhi — APPROVED** [review]

File: action/protocol/protocol.go [hunk 1]
```diff
@@ -78,6 +79,9 @@ type ActionHandler interface {
 	Handle(context.Context, action.Action, StateManager) (*action.Receipt, error)
 }
 
+// DepositGasAndBurn deposits gas to rewarding pool and burns baseFee
+type DepositGasAndBurn func(context.Context, StateManager, *big.Int, *big.Int) ([]*action.TransactionLog, error)
```
Comment (envestcc) [inline action/protocol/protocol.go:83]: adding `basefee` param as option seems simpler, not cause many changes in test

File: action/protocol/rewarding/fund.go [hunk 2]
```diff
@@ -146,3 +219,29 @@ func DepositGas(ctx context.Context, sm protocol.StateManager, amount *big.Int)
 	}
 	return rp.Deposit(ctx, sm, amount, iotextypes.TransactionLogType_GAS_FEE)
 }
+
+// DepositGasAndBurn deposits gas into the rewarding fund and burns the base fee
+func DepositGasAndBurn(ctx context.Context, sm protocol.StateManager, amount, baseFee *big.Int) ([]*action.TransactionLog, error) {
```
Comment (envestcc) [inline action/protocol/rewarding/fund.go:224]: merge with original function?

File: action/protocol/rewarding/fund.go [hunk 3]
```diff
@@ -146,3 +219,29 @@ func DepositGas(ctx context.Context, sm protocol.StateManager, amount *big.Int)
...
+	// If the gas fee is 0, return immediately
+	if amount.Cmp(big.NewInt(0)) == 0 {
+		return nil, nil
```
Comment (envestcc) [inline action/protocol/rewarding/fund.go:227]: What kind of scenario is this? Returning an error seems more appropriate.

File: action/protocol/rewarding/fund.go [hunk 4]
```diff
@@ -146,3 +219,29 @@ func DepositGas(ctx context.Context, sm protocol.StateManager, amount *big.Int)
...
+	reg, ok := protocol.GetRegistry(ctx)
+	if !ok {
+		return nil, nil
+	}
+	rp := FindProtocol(reg)
+	if rp == nil {
+		return nil, nil
+	}
+	burn, _ := address.FromString(address.ZeroAddress)
+	return rp.depositAndBurn(ctx, sm, burn, amount, baseFee, iotextypes.TransactionLogType_GAS_FEE, iotextypes.TransactionLogType_NATIVE_TRANSFER)
```
Comment (envestcc) [inline action/protocol/rewarding/fund.go:242]: Is it appropriate to use TransactionLogType_NATIVE_TRANSFER to denote burning? Will it cause any misunderstandings (e.g. transfer basefee to other people)?

File: action/protocol/staking/protocol.go [hunk 5]
```diff
@@ -690,17 +691,23 @@ const (
...
 ) (*action.Receipt, error) {
-	actionCtx := protocol.MustGetActionCtx(ctx)
-	blkCtx := protocol.MustGetBlockCtx(ctx)
-	gasFee := big.NewInt(0).Mul(actionCtx.GasPrice, big.NewInt(0).SetUint64(gasToBeDeducted))
-	depositLog, err := p.helperCtx.DepositGas(ctx, sm, gasFee)
+	var (
+		actionCtx = protocol.MustGetActionCtx(ctx)
+		blkCtx    = protocol.MustGetBlockCtx(ctx)
+	)
+	gasFee, baseFee, err := protocol.SplitGas(ctx, act, gasToBeDeducted)
```
Comment (envestcc) [inline action/protocol/staking/protocol.go:706]: take care of multiple protocol handling(e.g. MigrateStake), `baseFee` maybe double-deducted

File: action/protocol/rewarding/fund.go (hunk 4 above)
Comment (dustinxie) [reply action/protocol/rewarding/fund.go]: in blockchain, burn = send to the zero address
we can discuss, maybe send the amount to a struct (like `fund`) in rewarding protocol to keep track of all burned fees. and maybe can use this part for inflation

File: action/protocol/staking/protocol.go (hunk 5 above)
Comment (dustinxie) [reply action/protocol/staking/protocol.go:705]: should be OK, the `gasToBeDeducted ` here is only for the second part of action
the gas of first part (create NFT) is handled in EVM protocol

**Review by envestcc — APPROVED** [review]

**Pull request merged**